pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,74 +16,151 @@
+	storage_alias,
+	traits::{Get, GetStorageVersion, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion},
+	Blake2_128Concat,
 };
 use sp_std::marker::PhantomData;
 
-pub struct LookupReverseIndexMigration<T>(PhantomData<T>);
-
-impl<T: Config> OnRuntimeUpgrade for LookupReverseIndexMigration<T> {
-	#[cfg(feature = "try-runtime")]
-	fn pre_upgrade() -> Result<(), &'static str> {
-		assert!(Pallet::<T>::on_chain_storage_version() < Pallet::<T>::current_storage_version());
-		assert_eq!(ConnectedAccounts::<T>::iter().count(), 0);
-
-		log::info!(
-			"👥  DID lookup pallet to {:?} passes PRE migrate checks ✅",
-			Pallet::<T>::current_storage_version()
-		);
-
-		Ok(())
+#[cfg(feature = "try-runtime")]
+use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::SaturatedConversion;
+
+#[storage_alias]
+type ConnectedDids<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, AccountIdOf<T>, ConnectionRecordOf<T>>;
+#[storage_alias]
+type ConnectedAccounts<T: Config> =
+	StorageDoubleMap<Pallet<T>, Blake2_128Concat, DidIdentifierOf<T>, Blake2_128Concat, AccountIdOf<T>, ()>;
+#[storage_alias]
+type TmpConnectedDids<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, LinkableAccountId, ConnectionRecordOf<T>>;
+#[storage_alias]
+type TmpConnectedAccounts<T: Config> =
+	StorageDoubleMap<Pallet<T>, Blake2_128Concat, DidIdentifierOf<T>, Blake2_128Concat, LinkableAccountId, ()>;
+
+// Inspired by frame_support::storage::migration::move_storage_from_pallet
+fn move_storage<P: PalletInfoAccess>(old_storage_name: &[u8], new_storage_name: &[u8]) {
+	let pallet_name = <P as PalletInfoAccess>::name();
+
+	let old_prefix = storage_prefix(pallet_name.as_bytes(), old_storage_name);
+	let new_prefix = storage_prefix(pallet_name.as_bytes(), new_storage_name);
+	move_prefix(&old_prefix, &new_prefix);
+
+	if let Some(value) = unhashed::get_raw(&old_prefix) {
+		unhashed::put_raw(&new_prefix, &value);
+		unhashed::kill(&old_prefix);
 	}
+}
 
+pub struct EthereumMigration<T>(PhantomData<T>);
+
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for EthereumMigration<T>
+where
+	T::AccountId: Into<LinkableAccountId>,
+{
 	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		// Account for the new storage version written below.
-		let initial_weight = T::DbWeight::get().writes(1);
-
-		// Origin was disabled, so there cannot be any existing links. But we check just
-		// to be sure.
-		let total_weight: Weight =
-			ConnectedDids::<T>::iter().fold(initial_weight, |total_weight, (account, record)| {
-				ConnectedAccounts::<T>::insert(record.did, account, ());
-				// One read for the `ConnectedDids` entry, one write for the new
-				// `ConnectedAccounts` entry.
-				total_weight.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+		if Pallet::<T>::current_storage_version() == StorageVersion::new(3) {
+			// already on version 3
+			<T as frame_system::Config>::DbWeight::get().reads_writes(1, 0)
```

Comment:
never understood why rust removed the `return`. It's that much less readable IMHO

Reply:
Is it removed? You can still return.

Reply:
yeah but this is the last statement in the function. So this will be return. You can't take it out of the if branch. But you don't see that this is the last statement in the function. You have to guess that it is.